README.md
# aws-assume-role-lib
**Assumed role session chaining (with credential refreshing) for boto3**

The typical way to use boto3 when programmatically assuming a role is to explicitly call [`sts.AssumeRole`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts.html#STS.Client.assume_role) and use the returned credentials to create a new [`boto3.Session`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html) or client.
It looks like this mess of code:

```python
role_arn = "arn:aws:iam::123456789012:role/MyRole"
session = boto3.Session()

sts = session.client("sts")
response = sts.assume_role(
    RoleArn=role_arn,
    RoleSessionName="something_you_have_to_think_about"
)

credentials = response["Credentials"]

assumed_role_session = boto3.Session(
    aws_access_key_id=credentials["AccessKeyId"],
    aws_secret_access_key=credentials["SecretAccessKey"],
    aws_session_token=credentials["SessionToken"]
)

# use the session
print(assumed_role_session.client("sts").get_caller_identity())
```

This code is verbose, requires specifying a role session name even if you don't care what it is, and must explicitly handle credential expiration and refreshing if needed (in a Lambda function, this is typically handled by calling `AssumeRole` in every invocation).

With `aws-assume-role-lib`, all that collapses down to a single line. The assumed role session automatically refreshes expired credentials and generates a role session name if one is not provided.

```python
role_arn = "arn:aws:iam::123456789012:role/MyRole"
session = boto3.Session()

assumed_role_session = aws_assume_role_lib.assume_role(session, role_arn)

# use the session
print(assumed_role_session.client("sts").get_caller_identity())
```

In a Lambda function that needs to assume a role, you can create the assumed role session during initialization and use it for the lifetime of the execution environment, with `AssumeRole` calls only being made when necessary, not on every invocation.

Note that in `~/.aws/config`, [you have the option to have profiles that assume a role based on another profile](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-role.html), and this automatically handles refreshing expired credentials as well.

If you've only used `boto3.client()` and are not familiar with boto3 sessions, [here's an explainer](https://ben11kehoe.medium.com/boto3-sessions-and-why-you-should-use-them-9b094eb5ca8e).

# Installation

```bash
pip install --user aws-assume-role-lib
```

Or just add [`aws_assume_role_lib.py`](https://raw.githubusercontent.com/benkehoe/aws-assume-role-lib/stable/aws_assume_role_lib/aws_assume_role_lib.py) to your project.

View the release history [here](CHANGELOG.md).

# Usage

```python
import boto3
from aws_assume_role_lib import assume_role

# Get a session
session = boto3.Session()
# or with a profile:
# session = boto3.Session(profile_name="my-profile")

# Assume the session
assumed_role_session = assume_role(session, "arn:aws:iam::123456789012:role/MyRole")

# do stuff with the original credentials
print(session.client("sts").get_caller_identity()["Arn"])

# do stuff with the assumed role
print(assumed_role_session.client("sts").get_caller_identity()["Arn"])
```

In Lambda, initialize the sessions outside the handler, and `AssumeRole` will only get called when necessary, rather than on every invocation:
```python
import os
import boto3
from aws_assume_role_lib import assume_role, generate_lambda_session_name

# Get the Lambda session
SESSION = boto3.Session()

# Get the config
ROLE_ARN = os.environ["ROLE_ARN"]
ROLE_SESSION_NAME = generate_lambda_session_name() # see below for details

# Assume the session
ASSUMED_ROLE_SESSION = assume_role(SESSION, ROLE_ARN, RoleSessionName=ROLE_SESSION_NAME)

def handler(event, context):
    # do stuff with the Lambda role using SESSION
    print(SESSION.client("sts").get_caller_identity()["Arn"])

    # do stuff with the assumed role using ASSUMED_ROLE_SESSION
    print(ASSUMED_ROLE_SESSION.client("sts").get_caller_identity()["Arn"])
```
Learn more about the benefits of `aws-assume-role-lib` in Lambda functions in the [demo](lambda-demo/README.md).

# Interface

```
assume_role(
    # required arguments
    session:           boto3.Session,
    RoleArn:           str,

    *,
    # keyword-only arguments for AssumeRole
    RoleSessionName:   str                                    = None,
    PolicyArns:        Union[list[dict[str, str]], list[str]] = None,
    Policy:            Union[str, dict]                       = None,
    DurationSeconds:   Union[int, datetime.timedelta]         = None,
    Tags:              list[dict[str, str]]                   = None,
    TransitiveTagKeys: list[str]                              = None,
    ExternalId:        str                                    = None,
    SerialNumber:      str                                    = None,
    TokenCode:         str                                    = None,
    SourceIdentity:    str                                    = None,
    additional_kwargs: dict                                   = None,

    # keyword-only arguments for returned session
    region_name:       Union[str, bool]                       = None,

    # keyword-only arguments for assume_role() itself
    validate:          bool                                   = True,
    cache:             dict                                   = None,
)
```

`assume_role()` takes a session and a role ARN, and optionally [other keyword arguments for `sts.AssumeRole`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts.html#STS.Client.assume_role).

Unlike the `AssumeRole` API call itself, `RoleArn` is required, but `RoleSessionName` is not.
The `RoleSessionName` is set for you if it is not provided; it will use the [`SourceIdentity`](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_control-access_monitor.html) if that is provided, otherwise it will generated.
If you want this generated value for `RoleSessionName` when `SourceIdentity` is provided (the behavior in v2.8 and before), set `RoleSessionName` to the special value `aws_assume_role_lib.AUTOMATIC_ROLE_SESSION_NAME`.

Note that unlike the boto3 sts client method, you can provide the `Policy` parameter (the [inline session policy](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html#policies_session)) as a `dict` instead of as a serialized JSON string, `PolicyArns` as a list of ARNs, and `DurationSeconds` as a `datetime.timedelta` instead of as an integer.

By default, the session returned by `assume_role()` links its region configuration to the input session.
If you would like to set the region explicitly, pass it in the `region_name` parameter.

Note that if the parent session was created without a region passed in to the `Session` constructor, it has an implicit region, based on searching potential configuration locations.
This means that the region used by the session can change (for example, if you set or change `os.environ["AWS_DEFAULT_REGION"]`).
By default, the child session region is linked to the parent session, so if the parent session has an implicit region, or if the parent session's region is changed directly, they would both change.
If you would like to fix the child session region to be explicitly the current value, pass `region_name=True`.
If, for some reason, you have an explicit region set on the parent, and want the child to have implicit region config, pass `region_name=False`.

By default, `assume_role()` checks if the parameters are invalid.
Without this validation, errors for these issues are more confusingly raised when the child session is first used to make an API call (boto3 doesn't make the call to retrieve credentials until they are needed).
However, this incurs a small time penalty, so parameter validation can be disabled by passing `validate=False`.

If any new arguments are added to `AssumeRole` in the future and this library is not updated to allow them directly, they can be passed in as a dict via the `additional_kwargs` argument.

The parent session is available on the child session in the `assume_role_parent_session` property.
Note this property is added by this library; ordinary boto3 sessions do not have it.

# Patching boto3

You can make the `assume_role()` function available directly in boto3 by calling `patch_boto3()`.
This creates a `boto3.assume_role(RoleArn, ...)` function (note that it does not take a session, it uses the same default session as `boto3.client()`), and adds a `boto3.Session.assume_role()` method.
So usage for that looks like:

```python
import boto3
import aws_assume_role_lib
aws_assume_role_lib.patch_boto3()

assumed_role_session = boto3.assume_role("arn:aws:iam::123456789012:role/MyRole")
# the above is basically equivalent to:
# aws_assume_role_lib.assume_role(boto3.Session(), "arn:aws:iam::123456789012:role/MyRole")

session = boto3.Session(profile_name="my-profile")
assumed_role_session = session.assume_role("arn:aws:iam::123456789012:role/MyRole")
```

# Role session names for Lambda functions
Learn more about the benefits of `aws-assume-role-lib` in Lambda functions in the [demo](lambda-demo/README.md).

If you don't provide a role session name, but you provide a [`SourceIdentity`](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_control-access_monitor.html), this value is used for the role session name as well.
If `SourceIdentity` is not provided either, the underlying `botocore` library [generates one using a timestamp](https://github.com/boto/botocore/blob/c53072ec257ef47e2fc749c384a9488fd3f3e626/botocore/credentials.py#L730).
That's the best it can do, because it doesn't have any other context.

But in a Lambda function, we do have additional context, the Lambda function itself.
If you call `generate_lambda_session_name()` inside an instance of a Lambda function, it returns a session name that corresponds to the function instance, which you can use when assuming a role in the Lambda function (either with this library's `assume_role()` or any other method).
The purpose of this is to simplify tracing usage of the session back to the function instance.

The returned value is in one of the following forms, depending on the length of the values, to keep the session name within the maximum of 64 characters:
* `{function_name}`
* `{function_name}.{identifier}`
* `{function_name}.{function_version}.{identifier}`

The function version is never included if it is `$LATEST`.

The maximum role session name length is 64 characters. To ensure this, and
to provide at least 4 characters of the identifier when it is used, the
following rules apply, in order:
1. If the function name is longer than 59 characters, the session name is the truncated function name.

CHANGELOG.md
# Changelog

`aws-assume-role-lib` uses [monotonic versioning](http://blog.appliedcompscilab.com/monotonic_versioning_manifesto/).

## v2.10
* Fix CLI support
* Update boto3 version requirement for `SourceIdentity` validation

## v2.9
* Fixed `get_assumed_role_session_arn()`, which incorrectly used `iam` instead of `sts` in the ARN.
* If `RoleSessionName` is not provided but `SourceIdentity` is, the `SourceIdentity` value will be used for `RoleSessionName`, rather than the `botocore`-generated value.
    * Added special value `AUTOMATIC_ROLE_SESSION_NAME`; setting `RoleSessionName` to this will cause it to use pre-2.8 behavior (i.e., always use `botocore`-generated value).
* Updated for [PEP 561](https://www.python.org/dev/peps/pep-0561/) (typing) compliance. Link to single-file module has changed to https://raw.githubusercontent.com/benkehoe/aws-assume-role-lib/stable/aws_assume_role_lib/aws_assume_role_lib.py
    * Link to single-file module now uses `stable` branch, which will point to a release commit.

## v2.8
* If `region_name=None` (the default), the child session region is linked to the parent session region (not copied, which happens if `region_name=True`). If the parent session is using a config profile, this means the child session will have a region, rather than needing/using a separately-set region (e.g., via the `AWS_DEFAULT_REGION` environment variable). For safety, this release increments the [compatibility version](http://blog.appliedcompscilab.com/monotonic_versioning_manifesto/) to 2.
* Added [command line functionality](README.md#command-line-use).
* `PolicyArns` can be provided as a list of ARNs in addition to the verbose list-of-single-element-dicts required by the API.

## v1.7
* `generate_lambda_session_name()` now performs truncation to return a value that is always 64 characters or less.

## v1.6
* Ensured new functions get imported when doing `from aws_assume_role_lib import *`.

## v1.5
* Added `get_role_arn()` and `get_assumed_role_session_arn()` functions.

## v1.4
* Added support for the new [`SourceIdentity`](https://aws.amazon.com/blogs/security/how-to-relate-iam-role-activity-to-corporate-identity/) parameter.

Note that you don't need to wait for any future new parameters to be added to the library to start using them, because they can be passed in through the `additional_kwargs` parameter.
For `SourceIdentity`, before v1.4 this would have looked like including the parameter `additional_kwargs={'SourceIdentity': '...'}` to the `assume_role()` call.

## v1.3
* Added the `generate_lambda_session_name()` function.

## v1.2
* Added `region_name` parameter.
* Added `patch_boto3()` function.


aws_assume_role_lib/aws_assume_role_lib.py
# Copyright 2020 Ben Kehoe
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Assumed role session chaining (with credential refreshing) for boto3.

This library provides the ability to create boto3 Sessions that use a given
role, assuming that role from a parent session, in a way such that the child
session can refresh its credentials when they expire by implicitly calling
AssumeRole again."""

import typing
import json
import datetime
import os
import re
import numbers
import string
import random

import boto3
import botocore
import botocore.configprovider
import botocore.credentials
import botocore.exceptions
import botocore.session
import botocore.validate

__version__ = "2.10.0" # update here and pyproject.toml

__all__ = [
    "assume_role",
    "get_role_arn",
    "get_assumed_role_session_arn",
    "generate_lambda_session_name",
    "patch_boto3",
    "JSONFileCache",
    "AUTOMATIC_ROLE_SESSION_NAME",
]

class _ParentSessionProvider(botocore.configprovider.BaseProvider):
    def __init__(self, name: str, parent_session: botocore.session.Session):
        self.parent_session = parent_session
        self.name = name

    def provide(self):
        return self.parent_session.get_config_variable(self.name)

def _set_parent_session_provider(parent_session: botocore.session.Session, child_session: botocore.session.Session, name: str):
    # first check if an explicit value has been set in the child
    # otherwise get the value from the parent
    chain_provider = botocore.configprovider.ChainProvider([
        botocore.configprovider.InstanceVarProvider(name, child_session),
        _ParentSessionProvider(name, parent_session)
    ])
    config_store = child_session.get_component("config_store")
    config_store.set_config_provider(name, chain_provider)

# Force people to specify the path, which has a default in botocore
class JSONFileCache(botocore.credentials.JSONFileCache):
    """JSON file cache.

    This provides a dict like interface that stores JSON serializable
    objects.
    The objects are serialized to JSON and stored in a file.  These
    values can be retrieved at a later time.
    """

    def __init__(self, dir_path):
        super().__init__(working_dir=dir_path)

def get_role_arn(
        account_id: typing.Union[str, int],
        role_name: str,
        path: str="",
        partition: str="aws"):
    """Get a correctly-formatted IAM role ARN.

    You can provide the path separately or as part of the role name."""
    if isinstance(account_id, numbers.Number):
        account_id = str(int(account_id))
    if isinstance(account_id, str) and len(account_id) < 12:
        account_id = account_id.rjust(12, "0")

    if "/" in role_name and path:
        raise ValueError("Path cannot be given in both role_name and path")

    if "/" in role_name:
        path, role_name = role_name.rsplit("/", 1)

    if path == "/":
        path = ""
    if path.startswith("/"):
        path = path[1:]
    if path and not path.endswith("/"):
        path = path + "/"

    return f"arn:{partition}:iam::{account_id}:role/{path}{role_name}"

def get_assumed_role_session_arn(
        account_id: typing.Union[str, int],
        role_name: str,
        role_session_name: str,
        partition: str="aws"):
    """Get a correctly-formatted IAM assumed role session ARN.

    Note these ARNs do not contain the role's path, if it has one.
    If you provide the role name with path, it will be stripped off."""
    if isinstance(account_id, numbers.Number):
        account_id = str(int(account_id))
    if isinstance(account_id, str) and len(account_id) < 12:
        account_id = account_id.rjust(12, "0")

    if "/" in role_name:
        role_name = role_name.rsplit("/", 1)[1]

    return f"arn:{partition}:sts::{account_id}:assumed-role/{role_name}/{role_session_name}"

AUTOMATIC_ROLE_SESSION_NAME = "AUTOMATIC_ROLE_SESSION_NAME_!#$%^&*()" # use invalid chars

def assume_role(session: boto3.Session, RoleArn: str, *,
        RoleSessionName: str=None,
        PolicyArns: typing.Union[typing.List[typing.Dict[str, str]], typing.List[str]]=None,
        Policy: typing.Union[str, typing.Dict]=None,
        DurationSeconds: typing.Union[int, datetime.timedelta]=None,
        Tags: typing.List[typing.Dict[str, str]]=None,
        TransitiveTagKeys:typing.List[str]=None,
        ExternalId: str=None,
        SerialNumber: str=None,
        TokenCode: str=None,
        SourceIdentity: str=None,
        region_name: typing.Union[str, bool]=None,
        validate: bool=True,
        cache: dict=None,
        additional_kwargs: typing.Dict=None) -> boto3.Session:
    """Produce a boto3 Session using the given role, assumed using the input session.

    Unlike creating a session with the credentials returned directly
    by sts.AssumeRole, the returned session will refresh the credentials
    automatically when they expire by calling AssumeRole again.

    By default, the parameters are checked so that errors can be raised
    at this point, rather than more confusingly when the first call
    is made using the child session.
    This can be disabled by setting validate=False.

    The parent session is available on the child session
    in the property assume_role_parent_session.

    Additional arguments for AssumeRole, if any are added in the future,
    can be passed in additional_kwargs."""

    botocore_session = session._session

    if not RoleSessionName and SourceIdentity:
        RoleSessionName = SourceIdentity
    elif RoleSessionName == AUTOMATIC_ROLE_SESSION_NAME:
        RoleSessionName = None

    if PolicyArns:
        PolicyArns = [{"arn": value} if isinstance(value, str) else value for value in PolicyArns]

    if Policy is not None and not isinstance(Policy, str):
        Policy = json.dumps(Policy)

    if isinstance(DurationSeconds, datetime.timedelta):
        DurationSeconds = int(DurationSeconds.total_seconds())

    extra_args = {}
    if additional_kwargs:
        extra_args.update(additional_kwargs)

    for var_name in [
            "RoleSessionName",
            "PolicyArns",
            "Policy",
            "DurationSeconds",
            "Tags",
            "TransitiveTagKeys",
            "ExternalId",
            "SerialNumber",
            "TokenCode",
            "SourceIdentity"]:
        value = locals()[var_name]
        if value is not None:
            extra_args[var_name] = value

    credentials = botocore_session.get_credentials()
    if not credentials:
        raise botocore.exceptions.NoCredentialsError

    if validate:
        validate_args = extra_args.copy()
        validate_args["RoleArn"] = RoleArn
        if "RoleSessionName" not in validate_args:
            # this gets generated later if it's not present
            validate_args["RoleSessionName"] = "ToBeGenerated"
        shape = session.client("sts")._service_model.shape_for("AssumeRoleRequest")
        botocore.validate.validate_parameters(validate_args, shape)

    assume_role_provider = ProgrammaticAssumeRoleProvider(
        botocore_session.create_client,
        credentials,
        RoleArn,
        extra_args=extra_args,
        cache=cache,
    )

    assumed_role_botocore_session = botocore.session.Session()
    assumed_role_botocore_session.register_component(
        "credential_provider",
        botocore.credentials.CredentialResolver([assume_role_provider])
    )

    if region_name is True:
        region_name = session.region_name
    elif region_name is False:
        region_name = None
    elif region_name is None:
        try:
            _set_parent_session_provider(
                botocore_session,
                assumed_role_botocore_session,
                "region")
        except Exception as e:
            raise RuntimeError(
                "Unexpected breakage of botocore config API. " +
                "Fall back to setting region_name=True to use parent session region " +
                "or region=False to use implicit region.") from e

    session_kwargs = {
        "botocore_session": assumed_role_botocore_session,
        "region_name": region_name,
    }

    assumed_role_boto3_session = boto3.Session(**session_kwargs)

    assumed_role_boto3_session.assume_role_parent_session = session

    return assumed_role_boto3_session

def patch_boto3():
    """Add boto3.assume_role() and boto3.Session.assume_role().

    Each has the same interface as assume_role() except they do not take
    a session object as input."""
    setattr(boto3.Session, assume_role.__name__, assume_role)

    def wrapper(RoleArn: str, *,
            RoleSessionName: str=None,
            PolicyArns: typing.Union[typing.List[typing.Dict[str, str]], typing.List[str]]=None,
            Policy: typing.Union[str, typing.Dict]=None,
            DurationSeconds: typing.Union[int, datetime.timedelta]=None,
            Tags: typing.List[typing.Dict[str, str]]=None,
            TransitiveTagKeys:typing.List[str]=None,
            ExternalId: str=None,
            SerialNumber: str=None,
            TokenCode: str=None,
            SourceIdentity: str=None,
            region_name: typing.Union[str, bool]=None,
            validate: bool=True,
            cache: dict=None,
            additional_kwargs: typing.Dict=None) -> boto3.Session:
        """Produce a boto3 Session using the given role.

        Unlike creating a session with the credentials returned directly
        by sts.AssumeRole, the returned session will refresh the credentials
        automatically when they expire by calling AssumeRole again.

        By default, the parameters are checked so that errors can be raised
        at this point, rather than more confusingly when the first call
        is made using the child session.
        This can be disabled by setting validate=False.

        The parent session is available on the child session
        in the property assume_role_parent_session.

        Additional arguments for AssumeRole, if any are added in the future,
        can be passed in additional_kwargs."""
        session = boto3._get_default_session()
        return assume_role(session, RoleArn,
            RoleSessionName=RoleSessionName,
            PolicyArns=PolicyArns,
            Policy=Policy,
            DurationSeconds=DurationSeconds,
            Tags=Tags,
            TransitiveTagKeys=TransitiveTagKeys,
            ExternalId=ExternalId,
            SerialNumber=SerialNumber,
            TokenCode=TokenCode,
            SourceIdentity=SourceIdentity,
            region_name=region_name,
            validate=validate,
            cache=cache,
            additional_kwargs=additional_kwargs
        )
    wrapper.__name__ = assume_role.__name__
    setattr(boto3, assume_role.__name__, wrapper)

def generate_lambda_session_name(
        function_name: str=None,
        function_version: str=None,
        identifier: str=None):
    """For Lambda functions, generate a role session name that identifies the function.

    The returned value is in one of the following forms:
    {function_name}
    {function_name}.{identifier}
    {function_name}.{function_version}.{identifier}

    The function name must be retrievable from the AWS_LAMBDA_FUNCTION_NAME
    environment variable, or it must be provided.

    The function version is looked for in the AWS_LAMBDA_FUNCTION_VERSION
    environment variable by default. Function versions of $LATEST
    are treated the same as missing function versions.

    The identifier is taken from the log stream name in the
    AWS_LAMBDA_LOG_STREAM_NAME environment variable by default; if it is not
    provided and this cannot be found, it's a timestamp if the identifier can be
    at least 14 characters long (to provide for second-level precision),
    otherwise it is a random string.

    The maximum role session name length is 64 characters. To ensure this, and
    to provide at least 4 characters of the identifier when it is used, the
    following rules apply, in order:
    1. If the function name is longer than 59 characters, the session name is the
        truncated function name.
    2. If the function name plus the function version is longer than 59 characters,
        the session name is the function name plus the identifier, truncated.
    3. Otherwise, the session name is the function name plus the version (if one
        is found and not $LATEST) plus the identifier, truncated.
    """
    if not function_name:
        function_name = os.environ["AWS_LAMBDA_FUNCTION_NAME"]

    name_component = function_name

    if not function_version:
        function_version = os.environ.get("AWS_LAMBDA_FUNCTION_VERSION", "")

    if function_version and function_version != "$LATEST":
        version_component = "." + str(function_version)
    else:
        version_component = ""

    def _get_identifier(max_length):
        if identifier:
            return identifier
        # the execution environment has a unique ID, but it's not exposed directly
        # the log stream name (currently) includes it and looks like
        # 2020/01/31/[$LATEST]3893xmpl7fac4485b47bb75b671a283c
        log_stream_name = os.environ.get("AWS_LAMBDA_LOG_STREAM_NAME", "")
        match = re.search(r"\w+$", log_stream_name)
        if match:
            return match.group()[:max_length]
        elif max_length >= 14:
            # enough for second-level precision
            return datetime.datetime.utcnow().strftime("%Y%m%d%H%M%S%f")[:max_length]
        else:
            chars = string.ascii_lowercase + string.digits
            return ''.join(random.choice(chars) for _ in range(max_length))

    # truncate for max role session name length of 64
    if len(name_component) > 59:
        # don't bother with the identifier unless we can get
        # at least four characters of it
        value = name_component[:64]
    elif len(name_component) + len(version_component) > 59:
        # don't bother with the version if we can't get it
        max_length = 63 - len(name_component)
        identifier_component = "." + _get_identifier(max_length)
        value = f"{name_component}{identifier_component}"[:64]
    else:
        max_length = 63 - (len(name_component) + len(version_component))
        identifier_component = "." + _get_identifier(max_length)
        value = f"{name_component}{version_component}{identifier_component}"[:64]

    clean_value = re.sub(r"[^a-zA-Z0-9_=,.@-]+", "_", value)

    return clean_value

class ProgrammaticAssumeRoleProvider(botocore.credentials.CredentialProvider):
    METHOD = "assume-role"

    def __init__(self, client_creator, credential_loader, role_arn,
            extra_args=None, cache=None):
        self._client_creator = client_creator
        self._credential_loader = credential_loader
        self._role_arn = role_arn
        self._extra_args = extra_args
        self._fetcher = None
        if cache is None:
            cache = {}
        self._cache = cache

    def _get_fetcher(self):
        if not self._fetcher:
            self._fetcher = botocore.credentials.AssumeRoleCredentialFetcher(
                self._client_creator,
                self._credential_loader,
                self._role_arn,
                extra_args=self._extra_args,
                cache=self._cache
            )
        return self._fetcher

    def load(self):
        return botocore.credentials.DeferredRefreshableCredentials(
            self._get_fetcher().fetch_credentials,
            self.METHOD
        )

def main(arg_strs=None, exit=None):
    import argparse
    import sys
    import textwrap
    from datetime import datetime

    if not exit:
        exit = sys.exit

    TIME_FORMAT = '%Y-%m-%dT%H:%M:%SZ'

    def _json_dict_loader(input):
        try:
            data = json.loads(input)
            if not isinstance(data, dict):
                raise argparse.ArgumentTypeError("JSON value must be an object")
            return data
        except json.decoder.JSONDecodeError:
            raise argparse.ArgumentTypeError("must be a valid JSON object")

    def _dict_loader(input):
        try:
            data = json.loads(input)
            if not isinstance(data, dict):
                raise argparse.ArgumentTypeError("JSON value must be an object")
            return data
        except json.decoder.JSONDecodeError:
            pass
        return dict(v.split('=', 1) for v in input.split(','))

    def _list_loader(input):
        try:
            data = json.loads(input)
            if not isinstance(data, list):
                raise argparse.ArgumentTypeError("JSON value must be a list")
            return data
        except json.decoder.JSONDecodeError:
            pass
        return input.split(',')

    def _policy_arns_loader(input):
        try:
            data = json.loads(input)
            if not isinstance(data, (dict, list)):
                raise argparse.ArgumentTypeError("JSON value must be a list or object")
            return data
        except json.decoder.JSONDecodeError:
            pass
        return input.split(',')

    parser = argparse.ArgumentParser(
        prog="python -m aws_assume_role_lib",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description=textwrap.dedent("""\
    Assume the given role and print out the resulting credentials.

    For env var usage, export the output:
    export $( <invocation here> )

    In general, it's better to make profiles in ~/.aws/config for role assumption,
    which makes use of the AWS SDK's built-in support for role assumption.
    It also gets you automatic credential refreshing from the SDKs,
    unlike exporting them through this method.

    Example config profile for role assumption:

    [profile my-assumed-role]
    role_arn = arn:aws:iam::123456789012:role/MyRole
    # optional: role_session_name = MyRoleSessionName
    source_profile = profile-to-call-assume-role-with
    # or:
    # credential_source = Environment
    # credential_source = Ec2InstanceMetadata
    # credential_source = EcsContainer"""))

    parser.add_argument("--profile", help="the AWS config profile to use")
    parser.add_argument("RoleArn")

    group = parser.add_mutually_exclusive_group()
    group.add_argument("--env", action="store_const", const="env", dest="format", help="print as env vars (default)")
    group.add_argument("--json", action="store_const", const="json", dest="format", help="print credential_process-formatted JSON to stdout")
    parser.set_defaults(format="env")

    parser.add_argument("--RoleSessionName", metavar="ROLE_SESSION_NAME")
    parser.add_argument("--PolicyArns", type=_policy_arns_loader, metavar="POLICY_ARNS", help="Arn1,Arn2 JSON list or JSON object")
    parser.add_argument("--Policy", type=_json_dict_loader, help="JSON object")
    parser.add_argument("--DurationSeconds", type=int, metavar="DURATION_SECONDS")
    parser.add_argument("--Tags", type=_dict_loader, help="Key1=Value1,Key2=Value2 or JSON object")
    parser.add_argument("--TransitiveTagKeys", type=_list_loader, metavar="TRANSITIVE_TAG_KEYS", help="Key1,Key2 or JSON list")
    parser.add_argument("--ExternalId", metavar="EXTERNAL_ID")
    parser.add_argument("--SerialNumber", metavar="SERIAL_NUMBER")
    parser.add_argument("--TokenCode", metavar="TOKEN_CODE")
    parser.add_argument("--SourceIdentity", metavar="SOURCE_IDENTITY")
    parser.add_argument("--additional-kwargs", type=_json_dict_loader, help="JSON object")

    args = parser.parse_args(args=arg_strs)

    try:
        session = boto3.Session(profile_name=args.profile)
    except Exception as error:
        print("Unable to locate credentials: {}".format(error), file=sys.stderr)
        exit(3); return
    if not session.get_credentials():
        print("Unable to locate credentials", file=sys.stderr)
        exit(3); return

    assume_role_args = vars(args).copy()
    for field in ["profile", "format"]:
        assume_role_args.pop(field)
    assume_role_args.update({
        "session": session,
        "region_name": True, # doesn't really matter
    })

    try:
        assumed_role_session = assume_role(**assume_role_args)
        credentials = assumed_role_session.get_credentials()
        frozen_credentials = credentials.get_frozen_credentials()
    except botocore.exceptions.ClientError as error:
        code = error.response["Error"]["Code"]
        msg = error.response["Error"]["Message"]
        print("Error when assuming role [{}]: {}".format(code, msg), file=sys.stderr)
        exit(4); return
    except Exception as error:
        print("Error when assuming role: {}".format(error), file=sys.stderr)
        exit(4); return

    expiration = None
    if hasattr(credentials, '_expiry_time') and credentials._expiry_time:
        if isinstance(credentials._expiry_time, datetime):
            expiration = credentials._expiry_time.strftime(TIME_FORMAT)
        elif isinstance(credentials._expiry_time, str):
            expiration = credentials._expiry_time

    if args.format == "json":
        data = {
            "Version": 1,
            "AccessKeyId": frozen_credentials.access_key,
            "SecretAccessKey": frozen_credentials.secret_key,
            "SessionToken": frozen_credentials.token,
        }
        if expiration:
            data["Expiration"] = expiration

        print(json.dumps(data, indent=2))
    elif args.format == "env":
        lines = [
            "AWS_ACCESS_KEY_ID={}".format(frozen_credentials.access_key),
            "AWS_SECRET_ACCESS_KEY={}".format(frozen_credentials.secret_key),
            "AWS_SESSION_TOKEN={}".format(frozen_credentials.token),
        ]
        if expiration:
            lines.append("AWS_CREDENTIALS_EXPIRATION={}".format(expiration))
        print("\n".join(lines))
    else:
        print("Unexpected format {}".format(args.format), file=sys.stderr)
        exit(-1); return

if __name__ == "__main__":
    main()
